### PR TITLE
Fix the issue that functions that rely on `tipb::FieldType` may produce incorrect results

### LINKED ISSUE #8482: Functions that rely on `tipb::FieldType` may produce incorrect results
## Bug Report

such as https://github.com/pingcap/tidb/issues/49267

- cast as json
- tidb cast
- json_unquote

The reason is that `FieldType` does not affect the column name of the function, so functions with different `FieldTypes` will be reused.

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve?

Issue Number: close #8482

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Flash/Coprocessor/DAGUtils.cpp [hunk 1]
```diff
@@ -1417,6 +1421,11 @@ String genFuncString(const String & func_name, const Names & argument_names, con
             buf.append("_0");
     }
     buf.append(" ");
+    for (const auto * field_type : field_types)
+    {
+        if (field_type)
+            buf.append(field_type->DebugString());
```
[the remark this answers is not in the document]
Comment (SeaRise) [reply dbms/src/Flash/Coprocessor/DAGUtils.cpp]: good point, use `field_type->flag()` instead.

[the remark this answers is not in the document]
Comment (SeaRise) [reply dbms/src/Flash/Coprocessor/DAGUtils.cpp]: sorry, use `buffer.fmtAppend("{}|{}", field_type->flag(), field_type->flen());` instead

**Review by windtalker — APPROVED** [review]
LGTM

**Pull request merged**